Ensure usage of WriteOperation is thread-safe

szewai opened the pull request:
#### 723a649a68dfcb82c8f983012f0c698aaf58756d
<pre>
Ensure usage of WriteOperation is thread-safe
<a href="https://bugs.webkit.org/show_bug.cgi?id=282631">https://bugs.webkit.org/show_bug.cgi?id=282631</a>
<a href="https://rdar.apple.com/139301715">rdar://139301715</a>

Reviewed by Ben Nham.

NetworkCache::Storage currently accesses WriteOperation on different threads; see Storage::dispatchWriteOperation for
an example. However, WriteOperation is not thread-safe: it does not protect its members. To fix this issue, this patch
ensures NetworkCache::Storage only accesses WriteOperation on the main thread. The background threads gets an isolated
copy of WriteOperation&apos;s record and its identifier for performing activities. When background threads finish their job,
they pass back the identifier to main thread; then Storage can find the WriteOperation by identifier and finish it on
main thread.

* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
(WebKit::NetworkCache::Storage::WriteOperation::WriteOperation):
(WebKit::NetworkCache::Storage::WriteOperation::~WriteOperation):
(WebKit::NetworkCache::Storage::WriteOperation::identifier const):
(WebKit::NetworkCache::Storage::WriteOperation::WTF_REQUIRES_CAPABILITY):
(WebKit::NetworkCache::Storage::WriteOperation::invokeMappedBodyHandler):
(WebKit::NetworkCache::Storage::WriteOperation::invokeCompletionHandler):
(WebKit::NetworkCache::Storage::storeBodyAsBlob):
(WebKit::NetworkCache::Storage::removeFromPendingWriteOperations):
(WebKit::NetworkCache::retrieveFromMemory):
(WebKit::NetworkCache::Storage::dispatchWriteOperation):
(WebKit::NetworkCache::Storage::addWriteOperationActivity):
(WebKit::NetworkCache::Storage::removeWriteOperationActivity):
(WebKit::NetworkCache::Storage::finishWriteOperationActivity):
(WebKit::NetworkCache::Storage::retrieve):
(WebKit::NetworkCache::Storage::store):
(WebKit::NetworkCache::Storage::finishWriteOperation): Deleted.
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.h:
(WebKit::NetworkCache::Storage::Record::isolatedCopy):

Canonical link: <a href="https://commits.webkit.org/286261@main">https://commits.webkit.org/286261@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fa2f15433c7eac0f309fce353a97eb6a2d05c8a3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75334 "Passed style check") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28185 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79806 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26599 "Built successfully") 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63917 "Build is being retried. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (retry)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2566 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/59127 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/17359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78401 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/63917 "Build is being retried. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (retry)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64730 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39489 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/63917 "Build is being retried. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (retry)") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24926 "Built successfully") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/63917 "Build is being retried. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (retry)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22563 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81291 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2674 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1680 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/67376 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2825 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64717 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66653 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16595 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10606 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8773 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2631 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5452 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2656 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3591 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2663 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.h:
```diff
@@ -198,8 +201,10 @@ class Storage : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Storage,
+    HashCountedSet<uint64_t, AlreadyHashed> m_writeOperationActivities WTF_GUARDED_BY_LOCK(m_activitiesLock);
     Deque<std::unique_ptr<WriteOperation>> m_pendingWriteOperations;
-    HashSet<std::unique_ptr<WriteOperation>> m_activeWriteOperations;
+    HashMap<uint64_t, std::unique_ptr<WriteOperation>> m_activeWriteOperations;
```
I'm curious. Don't we prefer to use ObjectIdentifiers for identifiers? Additionally since we are using the identifier as a key to the HashMap and don't want to use a value that would corrupt the HashMap.

szewai replied:
I think we do if the identifier is used in different classes and especially across processes (e.g. included in IPC message). In this case the identifier is only used internally in `NetworkCache::Storage`.

RupinMittal replied:
Ah I see. Wouldn't it be better to use ObjectIdentifier here as well for the sake of consistency?

szewai replied:
What consistency? I mean the identifier is used only inside `NetworkCache::Storage` so it does not need to be `ObjectIdentifier`

RupinMittal replied:
Oh I just meant consistency with other identifiers in the code base.

cdumez replied:
> I think we do if the identifier is used in different classes and especially across processes

Our safer C++ guidelines certainly do not make any such distinction. Any identifier should be an ObjectIdentifier.

szewai replied:
ok I was still under the impression that is only needed for IPC code and saw ReadOperation identifier in the same file uses uint64_t, but it seems the latest guideline is for all code.
Updated.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -69,6 +69,12 @@ static uint64_t nextReadOperationOrdinal()
     return ++ordinal;
 }
 
+static uint64_t nextWriteOperationIdentifier()
```
We should be using ObjectIdentifier in new code.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -135,27 +135,61 @@ bool Storage::ReadOperation::finish()
 
+Storage::WriteOperationIdentifier Storage::WriteOperation::identifier() const
+{
+    return m_identifier;
```
This is one line, should be inlined above.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -135,27 +135,61 @@ bool Storage::ReadOperation::finish()
+
+Storage::Record Storage::WriteOperation::record() const
+{
+    return m_record;
```
This is one line, should be inlined above.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -135,27 +135,61 @@ bool Storage::ReadOperation::finish()
+    return m_identifier;
+}
+
+Storage::Record Storage::WriteOperation::record() const
```
Seems this should return a `const Storage::Record&`

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -885,54 +929,65 @@ bool Storage::shouldStoreBodyAsBlob(const Data& bodyData)
-    if (writeOperation.completionHandler)
-        writeOperation.completionHandler(error);
+    auto* writeOperation = m_activeWriteOperations.get(identifier);
+    ASSERT(writeOperation);
```
`RELEASE_ASSERT()`

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -885,54 +929,65 @@ bool Storage::shouldStoreBodyAsBlob(const Data& bodyData)
         auto channel = IOChannel::open(WTFMove(recordPath), IOChannel::Type::Create);
         size_t recordSize = recordData.size();
-        channel->write(0, recordData, WorkQueue::main(), [this, &writeOperation, recordSize](int error) {
+        channel->write(0, recordData, WorkQueue::main(), [this, identifier, recordSize](int error) {
```
Not new to this patch but why is is safe to capture `this` here?

szewai replied:
`WriteOperation` has a Ref to `NetworkCache::Storage`, so `this` will not be gone until the `CompletionHandler` if called.

cdumez replied:
And yet I pointed out a place where we need a `protectedThis` somehow. I don't think we should rely on such details for lifetime in 2024. Lifetime management should be clearer in modern code.

szewai replied:
Yes, I said I agree it's better (https://github.com/WebKit/WebKit/pull/36208#discussion_r1831408769). I was just replying comments in order and answering "why it is safe".

cdumez replied:
Yes, I read top to bottom :)

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -622,27 +656,31 @@ static Data encodeRecordMetaData(const RecordMetaData& metaData)
-    ++writeOperation.activeCount;
+    addWriteOperationActivity(writeOperationIdentifier);
+
+    RunLoop::main().dispatch([this, blob = WTFMove(blob), writeOperationIdentifier] {
```
Why is it safe to capture `this` here?

szewai replied:
`WriteOperation` has a Ref to `NetworkCache::Storage`

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -885,54 +929,65 @@ bool Storage::shouldStoreBodyAsBlob(const Data& bodyData)
 
-    m_activeWriteOperations.remove(&writeOperation);
+    // WriteOperation may hold last reference to this.
+    Ref protectedThis { *this };
```
This really goes against our modern patterns. The call site should be holding a smart pointer to this object instead.

szewai replied:
Good point; it's better let the task captures a strong Ref instead. And we don't need the `WriteOperation` to hold strong ref in that case.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -135,27 +135,61 @@ bool Storage::ReadOperation::finish()
+        , m_record(record)
+        , m_mappedBodyHandler(WTFMove(mappedBodyHandler))
+        , m_completionHandler(WTFMove(completionHandler))
+    {
```
Can we `ASSERT(isMainRunLoop());` ?

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -135,27 +135,61 @@ bool Storage::ReadOperation::finish()
-    CompletionHandler<void(int)> completionHandler;
+    ~WriteOperation()
+    {
+        if (m_completionHandler)
```
Can we `ASSERT(isMainRunLoop());` ?

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -885,54 +929,65 @@ bool Storage::shouldStoreBodyAsBlob(const Data& bodyData)
-        auto recordPath = recordPathForKey(writeOperation.record.key);
+    addToRecordFilter(record.key);
 
+    backgroundIOQueue().dispatch([this, identifier, record = crossThreadCopy(record)] {
```
Why is it safe to capture `this` here?

szewai replied:
`WriteOperation` has a Ref to `NetworkCache::Storage`

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.h:
```diff
@@ -198,8 +204,10 @@ class Storage : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Storage,
     WebCore::Timer m_readOperationTimeoutTimer;
 
+    Lock m_activitiesLock;
+    HashCountedSet<WriteOperationIdentifier> m_writeOperationActivities WTF_GUARDED_BY_LOCK(m_activitiesLock);
```
What is this used for? I see us adding and removing from it but didn't see any other usage?

szewai replied:
`remove()` returns true if count drops to 0 and item is removed.

bnham commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -885,54 +917,63 @@ bool Storage::shouldStoreBodyAsBlob(const Data& bodyData)
+    auto identifier = writeOperation.identifier();
+    m_activeWriteOperations.add(identifier, WTFMove(writeOperationPtr));
 
+    auto record = writeOperation.record();
```
This is a copy as well as the copy in the capture list of the lambda at L930. We also have the original copy of the write operation stored in the `m_activeWriteOperations` array. Seems like this could lead to some duplication of Blob, although perhaps this is not too bad if the blobs are backed by memory mapped files.

It seems like we can definitely get rid of the copy in the lambda capture list (e.g. `record = crossThreadCopy(WTFMove(record))`). I'm not sure about the other copy in `m_activeWriteOperations` without more surgery.

bnham replied:
Or maybe just `const auto& record = writeOperation.record()` here and leave the copy in the lambda below alone.

bnham replied:
Although for this to do anything I guess you would also need to implement `Record::isolatedCopy() &&` which doesn't seem to currently be implemented.

szewai replied:
Ok, added `Record::isolatedCopy() &&`.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -885,54 +917,63 @@ bool Storage::shouldStoreBodyAsBlob(const Data& bodyData)
-        auto recordPath = recordPathForKey(writeOperation.record.key);
+    addToRecordFilter(record.key);
 
+    backgroundIOQueue().dispatch([this, protectedThis = Ref { *this }, identifier, record = crossThreadCopy(record)]() mutable {
```
I agree with Ben, it would be nice to use `crossThreadCopy(WTFMove(record))` here.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -885,54 +917,63 @@ bool Storage::shouldStoreBodyAsBlob(const Data& bodyData)
+bool Storage::removeWriteOperationActivity(WriteOperationIdentifier identifier)
+{
+    Locker locker { m_activitiesLock };
+    ASSERT(m_writeOperationActivities.contains(identifier));
```
This seems fishy. Seems this function is always expected to return true, so why are we returning a boolean at all? And if we don't need that boolean, then why do we need `m_writeOperationActivities` at all?

szewai replied:
m_writeOperationActivities is a `HashCountedSet`, so it tracks of WriteOperation with activity and the count of activities.
`m_writeOperationActivities.contains(identifier)` is to ensure the WriteOperation has activity.
`m_writeOperationActivities.remove(identifier);`  is decrementing count by 1. It returns true if the count drops to 0.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -885,54 +917,63 @@ bool Storage::shouldStoreBodyAsBlob(const Data& bodyData)
 
-    if (writeOperation.completionHandler)
-        writeOperation.completionHandler(error);
+    auto* writeOperation = m_activeWriteOperations.get(identifier);
```
Any reason we cannot use a `take` here instead of a get and remove?

szewai replied:
That should work too; will update.

bnham commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -885,54 +917,63 @@ bool Storage::shouldStoreBodyAsBlob(const Data& bodyData)
+    addToRecordFilter(record.key);
 
+    backgroundIOQueue().dispatch([this, protectedThis = Ref { *this }, identifier, record = crossThreadCopy(record)]() mutable {
+        auto recordDirectorPath = recordDirectoryPathForKey(record.key);
```
nit: `recordDirectoryPath`

szewai committed:
Ensure usage of WriteOperation is thread-safe
https://bugs.webkit.org/show_bug.cgi?id=282631
rdar://139301715

Reviewed by Ben Nham.

NetworkCache::Storage currently accesses WriteOperation on different threads; see Storage::dispatchWriteOperation for
an example. However, WriteOperation is not thread-safe: it does not protect its members. To fix this issue, this patch
ensures NetworkCache::Storage only accesses WriteOperation on the main thread. The background threads gets an isolated
copy of WriteOperation's record and its identifier for performing activities. When background threads finish their job,
they pass back the identifier to main thread; then Storage can find the WriteOperation by identifier and finish it on
main thread.

* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
(WebKit::NetworkCache::Storage::WriteOperation::WriteOperation):
(WebKit::NetworkCache::Storage::WriteOperation::~WriteOperation):
(WebKit::NetworkCache::Storage::WriteOperation::identifier const):
(WebKit::NetworkCache::Storage::WriteOperation::WTF_REQUIRES_CAPABILITY):
(WebKit::NetworkCache::Storage::WriteOperation::invokeMappedBodyHandler):
(WebKit::NetworkCache::Storage::WriteOperation::invokeCompletionHandler):
(WebKit::NetworkCache::Storage::storeBodyAsBlob):
(WebKit::NetworkCache::Storage::removeFromPendingWriteOperations):
(WebKit::NetworkCache::retrieveFromMemory):
(WebKit::NetworkCache::Storage::dispatchWriteOperation):
(WebKit::NetworkCache::Storage::addWriteOperationActivity):
(WebKit::NetworkCache::Storage::removeWriteOperationActivity):
(WebKit::NetworkCache::Storage::finishWriteOperationActivity):
(WebKit::NetworkCache::Storage::retrieve):
(WebKit::NetworkCache::Storage::store):
(WebKit::NetworkCache::Storage::finishWriteOperation): Deleted.
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.h:
(WebKit::NetworkCache::Storage::Record::isolatedCopy):

Canonical link: https://commits.webkit.org/286261@main
